CRM/PivotReport/AbstractData.php

```diff
@@ -0,0 +1,526 @@
+        $multiValuesOffset = 0;
+      } else {
+        $result[] = array_values($row);
+      }
+      $i = count($result);
+
+      if ($i === self::ROWS_MULTIVALUES_LIMIT) {
+        break;
+      }
+
+      unset($this->multiValues[$key]);
+
+      $totalOffset++;
+    }
+
+    return array(
+      'info' => array(
+        'index' => $index,
+        'nextOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $totalOffset : $totalOffset + 1,
+        'multiValuesOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $multiValuesRows['info']['multiValuesOffset'] : 0,
+        'multiValuesTotal' => !empty($multiValuesRows['info']['multiValuesTotal']) ? $multiValuesRows['info']['multiValuesTotal'] : 0,
+      ),
+      'data' => $result,
+    );
+  }
+
+  /**
+   * Returns an array containing set of rows which are built basing on given $row
+   * and $fields array with indexes of multi values of the $row.
+   *
+   * @param array $row
+   *   A single Entity row
+   * @param array $fields
+   *   Array containing Entity multi value fields as keys and integer
+   *   indexes as values
+   * @param int $offset
+   *   Combination offset to start from
+   * @param int $limit
+   *   How many records can we generate?
+   *
+   * @return array
+   */
+  protected function populateMultiValuesRow(array $row, array $fields, $offset, $limit) {
+    $data = array();
+    $info = array(
+      'multiValuesTotal' => $this->getTotalCombinations($row, $fields),
+      'multiValuesOffset' => 0,
+    );
+    $found = true;
+    $i = 0;
+
+    while ($found) {
+      if ($i >= $offset) {
+        $rowResult = array();
+        foreach ($fields as $key => $index) {
+          $rowResult[$key] = $row[$key][$index];
+        }
+        $data[] = array_values(array_merge($row, $rowResult));
+      }
+      foreach ($fields as $key => $index) {
+        $found = false;
+        if ($index + 1 === count($row[$key])) {
+          $fields[$key] = 0;
+          continue;
+        }
+        $fields[$key]++;
+        $found = true;
+        break;
+      }
+      $i++;
+      if (($i - $offset === $limit) && $found) {
+        $info['multiValuesOffset'] = $i;
+        break;
+      }
+    }
+
+    return array(
+      'info' => $info,
+      'data' => $data,
+    );
+  }
+
+  /**
+   * Gets number of multivalues combinations for given Entity row.
+   *
+   * @param array $row
+   *   Entity row
+   * @param array $fields
+   *   Array containing all Entity fields
+   *
+   * @return int
+   */
+  protected function getTotalCombinations(array $row, array $fields) {
+    $combinations = 1;
+
+    foreach ($fields as $key => $value) {
+      if (!empty($row[$key]) && is_array($row[$key])) {
+        $combinations *= count($row[$key]);
+      }
+    }
+
+    return $combinations;
+  }
+
+  /**
+   * Returns a result of recursively parsed and formatted $data.
+   *
+   * @param mixed $data
+   *   Data element
```

Comment:
Is it possible for data to not be an array?

Reply:
Initially it's always an array (Activity row). But formatResult() process the value recursively going deeper into the particular row's field (and it's value - where it may be not an array at the end).